modules/standard/Map.chpl

```diff
@@ -0,0 +1,538 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This module contains the implementation of the map type which is a container
+  that stores key-value associations. 
+
+  Maps are not parallel safe by default, but can be made parallel safe by
+  setting the param formal `parSafe` to true in any map constructor. When
+  constructed from another map, the new map will inherit the parallel safety
+  mode of its originating map.
+*/
+module Map {
+
+  // Lock code lifted from modules/standard/Lists.chpl.
+  // Maybe they should be combined into a Locks module.
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  pragma "no doc"
+  class _LockWrapper {
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  record map {
+    param parSafe = false;
+    type keyType, valType;
```

Comment:
It seems that the current ordering of these fields doesn't permit one to declare a variable:

```chapel
var m: map(int, int);
```

which seems like it would be attractive.  Reordering the fields seems like it would support that pattern.  I didn't participate much in the design of the map API, so don't know whether there are advantages to the current approach that I'm not picking up on.

Reply:
I think this is a good suggestion. In rewriting things to use `map` instead of `array-as-map` I tripped over
```
var m: map(string, int);
```

several times before I learned to always use `new`.
```
var m = new map(string, int);
```

Reply:
@daviditen: Just checking in on this.  Seems like it should be an easy fix, but I haven't seen a PR or issue for it and didn't want it to get forgotten in this comment stream.  Have you had a chance to take a look?